### PR TITLE
docs - GDD - add information about specific type of deadlocks.

### PR DESCRIPTION
**Pull request by mkiyama**
Updated Admin Guide and and GUC in Ref Guice.

### CODE REVIEW

File: gpdb-doc/dita/admin_guide/dml.xml [hunk 1]
```diff
@@ -338,20 +338,33 @@
         deadlocks. This algorithm allows Greenplum Database to relax concurrent update and delete
         restrictions on heap tables. (Greenplum Database still employs table-level locking on AO/CO
         tables, restricting concurrent <codeph>UPDATE</codeph>, <codeph>DELETE</codeph>, and
-          <codeph>SELECT...FOR UPDATE</codeph> operations.) By default, the Global Deadlock Detector
-        is disabled. You can enable the Global Deadlock Detector by setting the server configuration
-        parameter <xref
+          <codeph>SELECT...FOR UPDATE</codeph> operations.) </p>
+      <p>By default, the Global Deadlock Detector is disabled and Greenplum Database executes the
+        concurrent update and delete operations on a heap table serially. You can enable the Global
```
Comment (dyozie) [inline gpdb-doc/dita/admin_guide/dml.xml]: You can enable the GDD -> You can enable these concurrent updates and the GDD

(Either here or in the next sentence, we should explicitly mention that enabling GDD also enables the concurrent operations on heap tables.)

File: gpdb-doc/dita/ref_guide/config_params/guc-list.xml [hunk 2]
```diff
@@ -3019,14 +3019,22 @@
...
-        these types of concurrent operations serially.</p>
-      <ul id="ul_trv_qqr_qgb">
-        <li>Concurrent update operations on a table distribution key</li>
-        <li>Concurrent update operations where a subplan in a query plan contains a motion
-          operation</li>
-      </ul>
-      <p>If the Global Deadlock Detector is enabled, Greenplum Database returns an error for the
-        concurrent operations if a deadlock is detected.</p>
+        concurrent update and delete operations on a heap table serially.</p>
+      <p>If the Global Deadlock Detector is enabled, it can determine that a deadlock exists, and
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/config_params/guc-list.xml]: If the GDD is enabled, it can determine that -> If the GDD is enabled, these concurrent updates are permitted and the GDD determines when

File: gpdb-doc/dita/ref_guide/config_params/guc-list.xml [hunk 3]
```diff
@@ -3019,14 +3019,22 @@
...
+        breaks the deadlock by cancelling one or more backend processes associated with the youngest
+        transaction(s) involved. </p>
+      <p>When the Global Deadlock Detector manages the following types of concurrent update or
+        delete transactions. Only one of the concurrent transactions will succeed. The other
+        transactions will fail with an error indicating concurrent updates on the same row is not
+          allowed.<ul id="ul_bxf_mpp_5gb">
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/config_params/guc-list.xml]: Same edit here as above.

**Comment by mkiyama** [comment]
@gaos1 
Is the GUC set classification correct?

Currently the GUC says this,
The parameter can be set for a database system, an individual database, or a session.

and the set classifications are:
master
session
restart

This is a link to the set classification definitions.
http://docs-gpdb-6x.cfapps.io/6-0/ref_guide/config_params/guc_config.html

**Comment by gaos1** [comment]
@mkiyama i didn't find the description in the doc about the guc scope. the guc gp_enable_global_deadlock_detector should be system scope and need cluster restart on change.

**Comment by mkiyama** [comment]
@gaos1 The information is in Table 1 on the page.
http://docs-gpdb-6x.cfapps.io/6-0/ref_guide/config_params/guc_config.html

File: gpdb-doc/dita/ref_guide/config_params/guc-list.xml [hunk 4]
```diff
@@ -3019,14 +3019,22 @@
...
+          <li>Concurrent update or delete transactions on the same row of a heap table where a
```
Comment (zhangjackey) [inline gpdb-doc/dita/ref_guide/config_params/guc-list.xml]: The sequence of concurrent update or delete is important, There are 4 cases:
* Transaction1 do **delete** first and then Transaction do **delete**, it will **not** raise an error.
* Transaction1 do **update** first and then Transaction do **delete**, it will raise an error.
* Transaction1 do **delete** first and then Transaction do **update**, it will **not** raise an error.
* Transaction1 do **update** first and then Transaction do **update**, it will raise an error.

In short, we can say that if update execute on ahead, and then an other transaction execute update or delete which process the same row and contains Motion operator, it will raise an error.

File: gpdb-doc/dita/admin_guide/dml.xml [hunk 5]
```diff
@@ -338,20 +338,33 @@
...
             type="section">gp_global_deadlock_detector_period</xref></codeph> server configuration
         parameter.</p>
+      <p>If the Global Deadlock Detector determines that deadlock exists, it breaks the deadlock by
+        cancelling one or more backend processes associated with the youngest transaction(s)
+        involved. </p>
+      <p>When the Global Deadlock Detector manages the following types of concurrent update or
+        delete transactions, only one of the concurrent transactions will succeed. The other
+        transactions will fail with an error indicating concurrent updates on the same row is not
+          allowed.<ul id="ul_bxf_mpp_5gb">
+          <li>Concurrent update or delete transactions on the same row of a heap table where a
```
[the remark this answers is not in the document]
Comment (mkiyama) [reply gpdb-doc/dita/admin_guide/dml.xml]: I'm leaving "concurrent" for emphasis.
It's a lot of "concurrents". I took one out of intro paragraph.

File: gpdb-doc/dita/ref_guide/config_params/guc-list.xml (hunk 4 above)
Comment (mkiyama) [reply gpdb-doc/dita/ref_guide/config_params/guc-list.xml]: @zhangjackey 
This is the updated description of concurrent transactions. Is this correct?
Thanks for your help.

-Concurrent transactions on the same row of a heap table where the first transaction is an update operation and a later transaction executes an update or delete and the query plan contains a motion operator.

-Concurrent update transactions on the same distribution key of a heap table that are executed by the Greenplum Database legacy query optimizer.

-Concurrent update transactions on the same row of a hash table that are executed by the GPORCA optimizer.

Comment (mkiyama) [reply gpdb-doc/dita/ref_guide/config_params/guc-list.xml]: @gaos1 @zhangjackey 
Are there any more technical changes required.

thanks.

File: gpdb-doc/dita/admin_guide/dml.xml [hunk 6]
```diff
@@ -338,20 +338,34 @@
         deadlocks. This algorithm allows Greenplum Database to relax concurrent update and delete
         restrictions on heap tables. (Greenplum Database still employs table-level locking on AO/CO
         tables, restricting concurrent <codeph>UPDATE</codeph>, <codeph>DELETE</codeph>, and
-          <codeph>SELECT...FOR UPDATE</codeph> operations.) By default, the Global Deadlock Detector
-        is disabled. You can enable the Global Deadlock Detector by setting the server configuration
-        parameter <xref
+          <codeph>SELECT...FOR UPDATE</codeph> operations.) </p>
+      <p>By default, the Global Deadlock Detector is disabled and Greenplum Database executes the
+        concurrent update and delete operations on a heap table serially. You can enable these
+        concurrent updates and the have the Global Deadlock Detector determine when a deadlock
```
Comment (lisakowen) [inline gpdb-doc/dita/admin_guide/dml.xml]: "the have" -> **have** ?

File: gpdb-doc/dita/admin_guide/dml.xml [hunk 7]
```diff
@@ -338,20 +338,34 @@
...
           href="../ref_guide/config_params/guc-list.xml#gp_enable_global_deadlock_detector"
             ><codeph>gp_enable_global_deadlock_detector</codeph></xref>. </p>
-      <p>If the Global Deadlock Detector determines that deadlock exists, it breaks the deadlock by
-        cancelling one or more backend processes associated with the youngest transaction(s)
-        involved.</p>
-      <p>The Global Deadlock Detector backend process is automatically started on the master host
-        when you start Greenplum Database. You configure the interval at which the Global Deadlock
-        Detector collects and analyzes lock waiting data via the <codeph><xref
+      <p>When the Global Deadlock Detector is enabled, the backend process is automatically started
+        on the master host when you start Greenplum Database. You configure the interval at which
```
Comment (lisakowen) [inline gpdb-doc/dita/admin_guide/dml.xml]: (re)start ?

File: gpdb-doc/dita/ref_guide/config_params/guc-list.xml [hunk 8]
```diff
@@ -3019,14 +3019,22 @@
...
+          id="ul_bxf_mpp_5gb">
+          <li>Concurrent transactions on the same row of a heap table where the first transaction is
+            an update operation and a later transaction executes an update or delete and the query
+            plan contains a motion operator.</li>
+          <li>Concurrent update transactions on the same distribution key of a heap table that are
+            executed by the Greenplum Database legacy query optimizer.</li>
+          <li>Concurrent update transactions on the same row of a hash table that are executed by
+            the GPORCA optimizer.</li>
+        </ul></p>
       <p>The parameter can be set for a database system, an individual database, or a session.</p>
```
Comment (lisakowen) [inline gpdb-doc/dita/ref_guide/config_params/guc-list.xml]: i don't think you can set this parameter for a database or a session.

File: gpdb-doc/dita/ref_guide/config_params/guc-list.xml [hunk 9]
```diff
@@ -3019,14 +3019,22 @@
...
-        <li>Concurrent update operations where a subplan in a query plan contains a motion
-          operation</li>
-      </ul>
-      <p>If the Global Deadlock Detector is enabled, Greenplum Database returns an error for the
-        concurrent operations if a deadlock is detected.</p>
+        concurrent update and delete operations on a heap table serially.</p>
+      <p>If the Global Deadlock Detector is enabled, it can determine that a deadlock exists, and
+        breaks the deadlock by cancelling one or more backend processes associated with the youngest
+        transaction(s) involved. </p>
+      <p>When the Global Deadlock Detector determines a deadlock exists for the following types of
```
Comment (lisakowen) [inline gpdb-doc/dita/ref_guide/config_params/guc-list.xml]: the first paragraph has a link to the GDD topic in the admin guide, which includes this bulleted list.  do we need to duplicate this information here?

File: gpdb-doc/dita/ref_guide/config_params/guc-list.xml (hunk 4 above)
Comment (gaos1) [reply gpdb-doc/dita/ref_guide/config_params/guc-list.xml]: @mkiyama the first one as shown below is specific to legacy planner, so need to add "by the Greenplum Database legacy query optimizer", similar as the second one.  All else looks good. Thanks. 

-Concurrent transactions on the same row of a heap table where the first transaction is an update operation and a later transaction executes an update or delete and the query plan contains a motion operator.

File: gpdb-doc/dita/admin_guide/dml.xml (hunk 7 above)
Comment (dyozie) [reply gpdb-doc/dita/admin_guide/dml.xml]: I think "start" is ok in this situation (it being a startup task, which would happen from a cold start or restart).
